### ISSUE TITLE
Less eagerly merge tuple unions to allow a typed mapping

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#9361]
I have found that pyright almost supports a typed key/value mapping and would like to request an enhancement so it's possible to implement something as a typed mapping similar to `TypedDict` but without the need for its special behaviour.

I believe that to make this possible the only change to pyright would be to less eagerly merge unions of tuple-pairs of values if the first is a literal string. However this could be applied to other literals as preferable to you.

Given the below:

```python
from typing import Any, Literal, LiteralString

type KVT = tuple[LiteralString, Any]

class Mapping[KvT: KVT]:
    def __getitem__[Kt: LiteralString, Vt, OtherT: KVT](
        self: Mapping[tuple[Kt, Vt] | OtherT],
        key: Kt
    ) -> Vt:
        ...

type IntKey = Literal["int_key"]
type IntKV = tuple[IntKey, int]
type StrKey = Literal["str_key"]
type StrKV = tuple[StrKey, str]

type MyMapping = Mapping[IntKV | StrKV]

class GetterFn[Tk: LiteralString]:
    @staticmethod
    def get[Tv, To: KVT](key: Tk, mapping: Mapping[tuple[Tk, Tv] | To]) -> Tv:
        ...

def use_get_fn(mapping: MyMapping):
    int_val = GetterFn[IntKey].get("int_key", mapping)
    reveal_type(int_val) # inferred type: int - correct
    str_val = GetterFn[StrKey].get("str_key", mapping)
    reveal_type(str_val) # inferred type: str - correct

    int_val = mapping["int_key"]
    reveal_type(int_val) # inferred type: int - correct
    str_val = mapping["str_key"] # type error - expects "int_key"
    reveal_type(str_val) # inferred type: int - incorrect
```
With an indirect getter type/function to  force pyright to use the string literal type, the above works to provide a typed mapping but is cumbersome. If pyright less eagerly merged the tuple union this would minimally allow a typed key/value mapping with distinct key/value types and wouldn't require the magic used in `TypedDict`.

I talk a bit more about this in the [Partial TypedDict](https://discuss.python.org/t/introduce-partial-for-typeddict/45176/65) discussion but this could in future be combined with `typing.Required` and `typing.NotRequired` to provide a notional partial typed mapping.

The above syntax may not be optimal but is functional.

I think this would give Python TypeScript's “super power”.

**Comment by alwaysmpe** [comment]
There's also a bug report in there (the inferred type in the `__getitem__` case should probably be `int | str`) but having it be `int` or `str` instead would be far cooler.

**Comment by erictraut** [comment]
The change you're suggesting would break other use cases, so it's not something I would do without a strong justification. 

Any time you rely on constraint solving behaviors involving unions of "raw" type variables, you're going to be on very shaky ground because they almost always allow for various (valid) constraint solutions, and the one that you end up with is dependent on the vagaries of the particular type checker's constraint solver implementation. Constraint solving behaviors are currently completely unspecified in the typing spec.

I'll note that mypy doesn't come close to accepting the above code, so it's not a pattern that could be relied upon generally unless/until more standardization is achieved.

**Comment by alwaysmpe** [comment]
Fair enough. Not the preferred answer but fully understand.

@erictraut  in that case there is also a bug where a union of type tuples containing a literal is not correctly merged, example:
```python
from typing import Any, Literal, LiteralString, reveal_type

type KVT = tuple[LiteralString, Any]

class Mapping[KvT]:
    def __getitem__[Kt: LiteralString, Vt, OtherT: KVT](
        self: Mapping[tuple[Kt, Vt] | OtherT],
        key: Kt
    ) -> Vt:
        ...

type IntKey = Literal["int_key"]
type IntKV = tuple[IntKey, int]
type StrKey = Literal["str_key"]
type StrKV = tuple[StrKey, str]

type MyMapping = Mapping[IntKV | StrKV]

def get_fn(mapping: MyMapping):
    int_val = mapping["int_key"]
    reveal_type(int_val) # inferred type: int - should really be int | str
    str_val = mapping["str_key"] # type error - expects "int_key" - should accept
    reveal_type(str_val) # inferred type: int - should really be int | str
```

And aside if of interest, I got the mapping type working somewhat as desired using an applicative monad, so a typed mapping is possible in the current type system without needing `TypedDict` (at least as currently implemented by pyright thanks to the hard work of yourself and others).
```python
from typing import Any, Literal, LiteralString

class Mapping[Kt: tuple[Any, Any]]:
    def __getitem__[GKt, Vt](
        self: Mapping[tuple[GKt, Vt]],
        key: GKt
    ) -> Vt:
        ...

class ApplyGet[Tk: LiteralString]:
    val: Tk
    def __init__(self, val: Tk):
        self.val = val

    def get[Vt, Vo: tuple[Any, Any]](self, mapping: Mapping[tuple[Tk, Vt] | Vo]) -> Vt:
        return mapping[self.val]

type IntKey = Literal["int_key"]
type IntKV = tuple[IntKey, int]
type StrKey = Literal["str_key"]
type ErrorStr = Literal["error_key"]
type StrKV = tuple[StrKey, str]
type MyMapping = Mapping[IntKV | StrKV]

def get_val(mapping: MyMapping):
    int_key: IntKey = "int_key"
    int_val = ApplyGet(int_key).get(mapping)
    raw_int_val = ApplyGet("int_key").get(mapping)
    reveal_type(int_val) # inferred type: int - correct
    reveal_type(raw_int_val) # inferred type: int - correct

    str_key: StrKey = "str_key"
    str_val = ApplyGet(str_key).get(mapping)
    raw_str_val = ApplyGet("str_key").get(mapping)
    reveal_type(str_val) # inferred type: str - correct
    reveal_type(raw_str_val) # inferred type: str - correct

    error_key: ErrorStr = "error_key"
    error_val = ApplyGet(error_key).get(mapping) # return type unknown error - correct
    raw_error_val = ApplyGet("error_key").get(mapping) # return type unknown error - correct
    reveal_type(error_val) # inferred type: unknown - correct
    reveal_type(raw_error_val) # inferred type: unknown - correct
```
I think this approach is a good way to go. As you mentioned, hardcoding the behaviour of operations like `Partial`, `Required`, `Pick`, and `Omit`  in python is unappealing and I think this approach could deliver as desired. I'll create a more fleshed out example and add it to the discussion for further consideration.
